### ISSUE TITLE
Error: theano.sandbox.cuda

### ISSUE THREAD

**Issue by ankitdhall** [Theano/Theano#3922]
I have installed CUDA and 64-bit Python using WinPython. My .theanorc.ini is placed in looks like the following:
[global]
floatX = float32
device = gpu
force_device = True

[nvcc]
flags = --use-local-env  --cl-version=2008
fastmath = True

[cuda]
## root = C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\bin

My nvcc.profile looks like the following:

TOP              = $(_HERE_)/..

NVVMIR_LIBRARY_DIR = $(TOP)/nvvm/libdevice

PATH            += $(TOP)/open64/bin;$(TOP)/nvvm/bin;$(_HERE_);$(TOP)/lib;

INCLUDES += "-I$(TOP)/include" "-I$(TOP)/include/cudart" $(_SPACE_)

LIBRARIES        =+ $(_SPACE_) "/LIBPATH:$(TOP)/lib/$(_WIN_PLATFORM_)"

CUDAFE_FLAGS    +=
## PTXAS_FLAGS     +=

After executing the script to ensure the GPU is working I get the following error on my Spyder IDE console:
4758 
4759 /**
4760  *
4761  \*  pattern is a permutation of [0, 1, ... self->nd-1] with the following twists:
4762  \*  - an element 'd' of the permutation can be dropped if CudaNdarray_HOST_DIMS(self)[d] == 1
4763  \*  - any number of '-1' elements can be in the pattern, and they will cause new ranks (with dim==1) to be inserted.
4764  *
4765  \*  For example, if CudaNdarray_HOST_DIMS(self) == [4, 5, 1, 6], and pattern = [0,3,-1,-1, 1], then CudaNdarray_HOST_DIMS(self) would be modified to become:
4766  \*     [4, 6, 1, 1, 5](we dropped the original dim[2]==1, and inserted two singleton dimensions with the -1s.
4767  */
4768 int
4769 CudaNdarray_dimshuffle%28CudaNdarray * self, unsigned int len, const int * pattern%29
4770 {
4771     //TODO: pass a workspace pointer to avoid the internal malloc
4772     int * newdims = %28int *%29malloc%28sizeof%28int%29 * %28len + len + self->nd%29%29; //we tack on the taken buffer here for speed of not having to malloc twice.
4773     int * newstrides = newdims + len;
4774     int * dims_taken = newstrides + len;
4775     if %28!newdims%29
4776     {
4777         PyErr_SetString%28PyExc_MemoryError,);
4778         return -1;
4779     }
4780     for (int i = 0; i < self->nd; ++i)
4781     {
4782         dims_taken[i] = 0;
4783     }
4784     for (int i = 0; i < len; ++i)
4785     {
4786         if (pattern[i] < 0)
4787         {
4788             newdims[i] = 1;
4789             newstrides[i] = 0;
4790         }
4791         else if(dims_taken[pattern[i]])
4792         {
4793             PyErr_Format(PyExc_ValueError, "Cudandarray_dimshuffle: invalid pattern for Cudandarray_dimshuffle. You used the dimensions %d multiple time",
4794                          pattern[i]);
4795             free(newdims);
4796             return -1;
4797         }
4798         else if (pattern[i]>= self->nd)
4799         {
4800             PyErr_Format(PyExc_ValueError, "Cudandarray_dimshuffle: invalid pattern for Cudandarray_dimshuffle. You asked for a dimensions that don't exist %d for a %d dims CudaNdarray",
4801                          pattern[i], self->nd);
4802             free(newdims);
4803             return -1;
4804         }
4805         else
4806         {
4807             newdims[i] = CudaNdarray_HOST_DIMS(self)[pattern[i]];
4808             newstrides[i] = CudaNdarray_HOST_STRIDES(self)[pattern[i]];
4809             dims_taken[pattern[i]] = 1;
4810         }
4811     }
4812     //Check if we dropped not broadcastable dims
4813     for (int i = 0; i < self->nd; ++i)
4814     {
4815         if (dims_taken[i]==0 && CudaNdarray_HOST_DIMS(self)[i]!=1)
4816         {
4817             PyErr_SetString(PyExc_ValueError, "Cudandarray_dimshuffle: You cannot drop a non-broadcastable dimension.");
4818             free(newdims);
4819             return -1;
4820         }
4821     }
4822     //swap this structure in for the one in self, and sync to the card
4823     if (CudaNdarray_set_nd(self, len))
4824     {
4825         free(newdims);
4826         return -1;
4827     }
4828     for (int i = 0; i < len; ++i)
4829     {
4830         CudaNdarray_set_dim(self, i, newdims[i]);
4831         CudaNdarray_set_stride(self, i, newstrides[i]);
4832     }
4833     if (cnda_copy_structure_to_device(self))
4834     {
4835         free(newdims);
4836         return -1;
4837     }
4838     free(newdims);
4839     return 0;
4840 }
4841 
4842 
4843 
4844 /**
4845  *
4846  \*  This is the function that bind to python.
4847  \*  See CudaNdarray_dimshuffle to call from C.
4848  \*  We use -1 to mean 'x' as in Tensor Dimshuffle.
4849  _/
4850 PyObject *
4851 CudaNdarray_Dimshuffle(PyObject_ _unused, PyObject\* args)
4852 {
4853     PyObject \* self = NULL;
4854     PyObject \* pattern_object = NULL;
4855     int \* pattern = NULL;
4856     PyObject \* rval = NULL;
4857     int success = -1;
4858     //const int \* dims = NULL;
4859 
4860     //args should consist of two python objects ("OO")
4861     if (! PyArg_ParseTuple(args, "OO", &self, &pattern_object))
4862         return NULL;
4863 
4864     if (!CudaNdarray_Check(self) )
4865     {
4866         PyErr_SetString(PyExc_TypeError, "First argument to cuda_ndarray.dimshuffle must be a CudaNdarray");
4867         return NULL;
4868     }
4869 
4870     //parse pattern_object into int \* pattern
4871 
4872     Py_ssize_t pattern_dim =  PyObject_Length(pattern_object);
4873 
4874     if (pattern_dim < 0)
4875     {
4876         PyErr_SetString(PyExc_TypeError, "Couldn't get length of third argument to cuda_ndarray.dimshuffle");
4877         return NULL;
4878     }
4879 
4880     pattern = (int _) malloc( pattern_dim \* sizeof(int));
4881 
4882     for (Py_ssize_t i = 0; i < pattern_dim; i++)
4883     {
4884         PyObject \* idx = PyLong_FromLong(i);
4885 
4886         if (idx == NULL)
4887         {
4888             PyErr_SetString(PyExc_Exception, "Couldn't make long object to loop over list/tuple");
4889             goto CudaNdarray_dimshuffle_fail;
4890         }
4891 
4892         long elem_value = 0;
4893 
4894         PyObject \* elem = PyObject_GetItem(pattern_object, idx);
4895 
4896         if (elem == NULL)
4897         {
4898             Py_XDECREF( elem);
4899             PyErr_SetString(PyExc_ValueError, "Third argument to dimshuffle must be list or tuple of integers");
4900             goto CudaNdarray_dimshuffle_fail;
4901         }
4902 
4903         elem_value = PyInt_AsLong(elem);
4904 
4905         if (elem_value == -1 && PyErr_Occurred() )
4906         {
4907             Py_XDECREF(elem);
4908             PyErr_SetString(PyExc_ValueError, "Third argument to dimshuffle must be list or tuple of integers");
4909             goto CudaNdarray_dimshuffle_fail;
4910         }
4911 
4912         pattern[i] = elem_value;
4913 
4914         Py_XDECREF( elem );
4915         Py_XDECREF( idx );
4916     }
4917 
4918     //allocate rval
4919     rval =  (PyObject *) CudaNdarray_View((CudaNdarray *) self);
4920 
4921     if (rval == NULL)
4922     {
4923         //CudaNdarray_New should have set the exception string
4924         goto CudaNdarray_dimshuffle_fail;
4925     }
4926 
4927 
4928     //printf("pattern_dim: %d\n",pattern_dim);
4929     //printf("pattern: %d %d\n",pattern[0],pattern[1]);
4930     //dims = CudaNdarray_HOST_DIMS( (CudaNdarray *) self);
4931     //printf("dims before: %d %d\n",dims[0],dims[1]);
4932 
4933     success = CudaNdarray_dimshuffle((CudaNdarray *) rval, pattern_dim, pattern);
4934 
4935     if (success != 0)
4936     {
4937         //Exception string should already be set by CudaNdarray_dimshuffle
4938         goto CudaNdarray_dimshuffle_fail;
4939     }
4940 
4941     free(pattern);
4942 
4943     return rval;
4944 
4945     CudaNdarray_dimshuffle_fail:
4946 
4947     if (pattern != NULL)
4948         free(pattern);
4949 
4950     Py_XDECREF(rval);
4951     return NULL;
4952 }
4953 
4954 
4955 int
4956 cnda_structure_size(int nd)
4957 {
4958     // dim0, dim1, ...
4959     // str0, str1, ...
4960     // log2(dim0), log2(dim1), ...
4961     return nd + nd + nd;
4962 }
4963 
4964 const int *
4965 CudaNdarray_HOST_DIMS(const CudaNdarray \* self)
4966 {
4967     return self->host_structure;
4968 }
4969 
4970 const int *
4971 CudaNdarray_HOST_STRIDES(const CudaNdarray \* self)
4972 {
4973     return self->host_structure + self->nd;
4974 }
4975 const int *
4976 CudaNdarray_HOST_LOG2DIMS(const CudaNdarray \* self)
4977 {
4978     return self->host_structure + 2_self->nd;
4979 }
4980 
4981 int
4982 CudaNdarray_EqualAndIgnore(CudaNdarray _cnda1, CudaNdarray *cnda2, int ignoreSync, int ignoreBase)
4983 {
4984     int verbose = 0;
4985 
4986     if (!ignoreSync && cnda1->dev_structure_fresh != cnda2->dev_structure_fresh)
4987     {
4988         if(verbose) fprintf(stdout, "CUDANDARRAY_EQUAL FAILED : 1\n");
4989         return 0;
4990     }
4991 
4992     if (cnda1->nd != cnda2->nd)
4993     {
4994         if(verbose) fprintf(stdout, "CUDANDARRAY_EQUAL FAILED : 2\n");
4995         return 0;
4996     }
4997 
4998     for (int i=0; i < 2_cnda1->nd; i++)
4999     {
5000         if (cnda1->host_structure[i] != cnda2->host_structure[i])
5001         {
5002             if(verbose)
5003                 fprintf(stdout, "CUDANDARRAY_EQUAL : host_structure : %d, %d, %d\n", i, cnda1->host_structure[i], cnda2->host_structure[i]);
5004             return 0;
5005         }
5006     }
5007 
5008     if (!ignoreBase && cnda1->base != cnda2->base)
5009     {
5010         if(verbose) fprintf(stdout, "CUDANDARRAY_EQUAL FAILED : 4");
5011         return 0;
5012     }
5013     else if (cnda1->data_allocated != cnda2->data_allocated)
5014     {
5015         if(verbose) fprintf(stdout, "CUDANDARRAY_EQUAL FAILED : 5");
5016         return 0;
5017     }
5018     else if (cnda1->data_allocated && cnda1->devdata != cnda2->devdata)
5019     {
5020         if(verbose) fprintf(stdout, "CUDANDARRAY_EQUAL FAILED : 6");
5021         // no need to check devdata if data is not allocated
5022         return 0;
5023     }
5024 
5025     return 1;
5026 }
5027 
5028 
5029 int
5030 CudaNdarray_Equal(CudaNdarray _cnda1, CudaNdarray *cnda2)
5031 {
5032     return CudaNdarray_EqualAndIgnore(cnda1, cnda2, 0, 0);
5033 }
5034 
5035 int
5036 cnda_copy_structure_to_device(const CudaNdarray \* self)
5037 {
5038     //If the device structure do not exists, create it.
5039     //We allocate it here as we do not need it often.
5040     //In fact, we need it so infrequently that we expect
5041     //that most object won't need it. Not allocating it
5042     //save a significant when creating object.
5043     //This speed up a benchmark by 8% with the gc.
5044     if (!self->dev_structure)
5045     {
5046         int struct_size = cnda_structure_size(self->nd);
5047         if (struct_size)
5048         {
5049             self->dev_structure = (int_)device_malloc(struct_size\* sizeof(int));
5050             if (NULL == self->dev_structure)
5051             {
5052                 return -1;
5053             }
5054         }
5055     }
5056     if (cublasSetVector(cnda_structure_size(self->nd),
5057                         sizeof(int),
5058                         self->host_structure,
5059                         1,
5060                         self->dev_structure,
5061                         1) != CUBLAS_STATUS_SUCCESS)
5062     {
5063         PyErr_SetString(PyExc_RuntimeError, "error copying structure to device memory");
5064         return -1;
5065     }
5066     self->dev_structure_fresh = 1;
5067     return 0;
5068 }
5069 
5070 const int *
5071 CudaNdarray_DEV_DIMS(const CudaNdarray \* self)
5072 {
5073     if (!self->dev_structure_fresh)
5074     {
5075         if (cnda_copy_structure_to_device(self))
5076             return NULL;
5077     }
5078     return self->dev_structure;
5079 }
5080 const int *
5081 CudaNdarray_DEV_STRIDES(const CudaNdarray \* self)
5082 {
5083     if (!self->dev_structure_fresh)
5084     {
5085         if (cnda_copy_structure_to_device(self))
5086             return NULL;
5087     }
5088     return self->dev_structure + self->nd;
5089 }
5090 const int *
5091 CudaNdarray_DEV_LOG2DIMS(const CudaNdarray \* self)
5092 {
5093     if (!self->dev_structure_fresh)
5094     {
5095         if (cnda_copy_structure_to_device(self))
5096             return NULL;
5097     }
5098     return self->dev_structure + 2_self->nd;
5099 }
5100 float *
5101 CudaNdarray_DEV_DATA(const CudaNdarray \* self)
5102 {
5103     return self->devdata;
5104 }
5105 
5106 /_*
5107  \* Return the number of elements in the ndarray (product of the dimensions)
5108  _/
5109 int
5110 CudaNdarray_SIZE(const CudaNdarray *self)
5111 {
5112     if (self->nd == -1) return 0;
5113     int size = 1;
5114     for (int i = 0; i < self->nd; ++i)
5115     {
5116         size *= CudaNdarray_HOST_DIMS(self)[i];
5117     }
5118     return size;
5119 }
5120 size_t
5121 CudaNdarray_SIZEt(const CudaNdarray *self)
5122 {
5123     if (self->nd == -1) return 0;
5124     size_t size = 1;
5125     for (int i = 0; i < self->nd; ++i)
5126     {
5127         size *= CudaNdarray_HOST_DIMS(self)[i];
5128     }
5129     return size;
5130 }
5131 
5132 PyObject *
5133 CudaNdarray_SIZE_Object(const CudaNdarray *self, void *closure)
5134 {
5135     return PyInt_FromLong(CudaNdarray_SIZE(self));
5136 }
5137 
5138 int CudaNdarray_set_device_data(CudaNdarray \* self, float \* data, const CudaNdarray \* base)
5139 {
5140     return CudaNdarray_set_device_data(self, data, (PyObject *) base);
5141 }
5142 
5143 PyObject \* CudaNdarray_IS_C_Contiguous(CudaNdarray \* self)
5144 {
5145     return PyBool_FromLong(CudaNdarray_is_c_contiguous(self));
5146 }
5147 
5148 int fprint_CudaNdarray(FILE \* fd, const CudaNdarray *self)
5149 {
5150     cudaError_t err = cudaGetLastError();
5151     if( cudaSuccess != err)
5152     {
5153         PyErr_Format(PyExc_RuntimeError,
5154                      "Cuda error: %s: %s.",
5155                      "fprint_CudaNdarray was called with an uncleared error",
5156                      cudaGetErrorString(err));
5157         return -1;
5158     }
5159     fprintf(fd, "CudaNdarray <%p, %p> nd=%i dev_structure_fresh=%d data_allocated=%d\n",
5160             self, self->devdata, self->nd, self->dev_structure_fresh, self->data_allocated);
5161     fprintf(fd, "\tHOST_DIMS:      ");
5162     for (int i = 0; i < self->nd; ++i)
5163     {
5164         fprintf(fd, "%i\t", CudaNdarray_HOST_DIMS(self)[i]);
5165     }
5166     fprintf(fd, "\n\tHOST_STRIDES: ");
5167     for (int i = 0; i < self->nd; ++i)
5168     {
5169         fprintf(fd, "%i\t", CudaNdarray_HOST_STRIDES(self)[i]);
5170     }
5171 
5172     if (self->dev_structure)
5173     {
5174         int data=0;
5175         fprintf(fd, "\n\tDEV_DIMS:      ");
5176         for (int i = 0; i < self->nd; ++i)
5177         {
5178             cublasGetVector(1, sizeof(int),
5179                             self->dev_structure+i, 1,
5180                             &data, 1);
5181             fprintf(fd, "%i\t", data);
5182         }
5183         fprintf(fd, "\n\tDEV_STRIDES: ");
5184         for (int i = 0; i < self->nd; ++i)
5185         {
5186             cublasGetVector(1, sizeof(int),
5187                             self->dev_structure + self->nd+i, 1,
5188                             &data, 1);
5189             fprintf(fd, "%i \t", data);
5190         }
5191         fprintf(fd, "\n");
5192     }
5193     else
5194     {
5195         fprintf(fd, "\n\tdev_structure not allocated\n");
5196     }
5197 
5198     err = cudaGetLastError();
5199     if( cudaSuccess != err)
5200     {
5201         PyErr_Format(PyExc_RuntimeError,
5202                      "Cuda error: %s: %s.",
5203                      "fprint_CudaNdarray",
5204                      cudaGetErrorString(err));
5205         return -1;
5206     }
5207     return 0;
5208 }
5209 
5210 
5211 int CudaNdarray_prep_output(CudaNdarray *_ arr, int nd,
5212                             const int \* dims, int fortran)
5213 {
5214     bool allocated = false;
5215     if (_arr == NULL)
5216     {
5217         // This allocates the metadata but not the data
5218         *arr = (CudaNdarray *) CudaNdarray_new_nd(nd);
5219         if (_arr == NULL)
5220             return -1;
5221         allocated = true;
5222     }
5223 
5224     if (CudaNdarray_alloc_contiguous(_arr, nd, dims, fortran))
5225     {
5226         if (allocated)
5227         {
5228             Py_DECREF(_arr);
5229             _arr = NULL;
5230         }
5231         return -1;
5232     }
5233     return 0;
5234 }
5235 
5236 
5237 /_
5238   Local Variables:
5239   mode:c++
5240   c-basic-offset:4
5241   c-file-style:"stroustrup"
5242   indent-tabs-mode:nil
5243   fill-column:79
5244   End:
5245 */
5246 // vim: filetype=cpp:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=79 :
#5247 

c:\program files\nvidia gpu computing toolkit\cuda\v7.5\include\host_config.h(191) : fatal error C1083: Cannot open include file: 'crtdefs.h': No such file or directory

ERROR (theano.sandbox.cuda): Failed to compile cuda_ndarray.cu: ('nvcc return status', 2, 'for cmd', 'nvcc -shared -O3 -Xlinker /DEBUG -D HAVE_ROUND -m64 -Xcompiler -DCUDA_NDARRAY_CUH=11b90075e2397c684f9dc0f7276eab8f,-D NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD -IC:\scisoft\WinPython-64bit-2.7.10.3\python-2.7.10.amd64\lib\site-packages\theano\sandbox\cuda -IC:\scisoft\WinPython-64bit-2.7.10.3\python-2.7.10.amd64\lib\site-packages\numpy\core\include -IC:\scisoft\WinPython-64bit-2.7.10.3\python-2.7.10.amd64\include -o C:\Users\Ankit\AppData\Local\Theano\compiledir_Windows-8-6.2.9200-Intel64_Family_6_Model_61_Stepping_4_GenuineIntel-2.7.10-64\cuda_ndarray\cuda_ndarray.pyd mod.cu -LC:\scisoft\WinPython-64bit-2.7.10.3\python-2.7.10.amd64\libs -LC:\scisoft\WinPython-64bit-2.7.10.3\python-2.7.10.amd64 -lpython27 -lcublas -lcudart')
[Elemwise{exp,no_inplace}(<TensorType(float64, vector)>)]
Looping 1000 times took 16.952000 seconds
Result is [ 1.23178032  1.61879341  1.52278065 ...,  2.20771815  2.29967753
  1.62323285]
Used the cpu

I am unable to find support on this issue, help would be really appreciated. What do you think went wrong?

Thank you.
Regards,
Ankit Dhall

**Comment by magic282** [comment]
Your VS version is too low to support CUDA 7.5. Update the VS to VS2013 will fix this.

**Comment by ankitdhall** [comment]
@magic282 yes, I will try to install a newer version.

**Comment by ankitdhall** [comment]
I have installed VS 2015 and the above errors aren't encountered but I am still unable to use the gpu. After running the script provided at http://deeplearning.net/software/theano/tutorial/using_gpu.html
I am still getting 'used the cpu'. I think the system doesn't recognize the configuration I have set. Although the .theanorc file has device=gpu, while executing print(theano.config.device) I get 'cpu'. Help would be appreciated. :)

**Comment by magic282** [comment]
@ankitdhall Don't know what happened, but just a reminder, VS 2015 is not supported by any CUDA versions. The latest supported VS is VS 2013 since the nvcc compiler needs the VC compiler. So VS 2010, 2012 and 2013 can be used here.

**Comment by ankitdhall** [comment]
@magic282 I have rolled back to Win 8.1 and now have VS 2013. I have placed the .theanorc file in C:\Users\Ankit\.theanorc. Also, while installing theano from the deeplearning website, I installed WinPython and they suggested to put the .theanorc file in the WinPython\Settings folder, which I did too.

But print(theano.config.device) still show cpu. Could there be a problem detecting the .theanorc file?
